narwhal/primary/src/core.rs

```diff
@@ -182,328 +180,308 @@ impl Core {
+                        // ensure liveness.
+                        return Ok(())
+                    }
+                },
+            }
+        }
     }
 
-    #[async_recursion]
-    #[instrument(level = "debug", skip_all, fields(vote_digest = ?vote.digest(), round = ?vote.round, from = ?vote.author))]
-    async fn process_vote(&mut self, vote: Vote) -> DagResult<()> {
-        debug!("Processing vote {:?}", vote);
-
-        // Add it to the votes' aggregator and try to make a new certificate.
-        if let Some(certificate) =
-            self.votes_aggregator
-                .append(vote, &self.committee, &self.current_header)?
-        {
-            debug!("Assembled {:?}", certificate);
-
-            // Broadcast the certificate.
-            let network_keys = self
-                .committee
-                .others_primaries(&self.name)
-                .into_iter()
-                .map(|(_, _, network_key)| network_key)
-                .collect();
-            let message = PrimaryMessage::Certificate(certificate.clone());
-            let handlers = self.network.broadcast(network_keys, &message).await;
-            self.cancel_handlers
-                .entry(certificate.round())
-                .or_insert_with(Vec::new)
-                .extend(handlers);
+    #[instrument(level = "debug", skip_all, fields(certificate_digest = ?certificate.digest()))]
+    async fn process_own_certificate(&mut self, certificate: Certificate) -> DagResult<()> {
+        self.metrics
+            .certificates_created
+            .with_label_values(&[&certificate.epoch().to_string()])
+            .inc();
+        self.metrics
+            .header_to_certificate_latency
+            .with_label_values(&[&certificate.epoch().to_string()])
+            .observe(
+                certificate
+                    .header
+                    .metadata
+                    .created_at
+                    .elapsed()
+                    .as_secs_f64(),
+            );
 
-            self.metrics
-                .certificates_created
-                .with_label_values(&[&certificate.epoch().to_string()])
-                .inc();
+        // Broadcast the certificate.
+        let network_keys = self
+            .committee
+            .others_primaries(&self.name)
+            .into_iter()
+            .map(|(_, _, network_key)| network_key)
+            .collect();
+        let tasks = self
+            .network
+            .broadcast(
+                network_keys,
+                &PrimaryMessage::Certificate(certificate.clone()),
+            )
+            .await;
+        self.background_tasks
+            .spawn(Self::send_certificates_while_current(
+                certificate.header.round,
+                tasks,
+                self.rx_narwhal_round_updates.clone(),
+            ));
+
+        // Process the new certificate.
+        match self.process_certificate_internal(certificate).await {
```

Comment:
I don't think so, there can't be errors and if there are we panic. I think we should arrange the code to be fastest in the optimistic case that the certificate succeeds as we expect. Remote nodes already have to guard against byzantine broadcasts.

Reply:
Agreed that this is not a case that should happen often (if ever). But a small counter point would be that it seems to make sense to wait till it is persisted in [local store ](https://github.com/MystenLabs/sui/blob/4b4a5a1828086f8197ca5156d0b01a3400e4dde0/narwhal/primary/src/core.rs#L550)because this node is supposed to be the source of truth for the cert it is broadcasting and if something happens before the write to store completes then we could be in a bad state or have some wasted effort across nodes.